lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2228,8 +2228,58 @@ namespace Js
         }
 
         ScriptContext* scriptContext = pThis->type->GetScriptContext();
-
         ApiError error = ApiError::NoError;
+        const char16 *pThisSz = pThis->GetSz();
+        charcount_t pThisLength = pThis->GetLength();
+
+        if (useInvariant)
+        {
+            bool isAscii = true;
+            for (charcount_t i = 0; i < pThisLength; i++)
+            {
+                if (pThisSz[i] >= 0x80)
+                {
+                    isAscii = false;
+                    break;
+                }
+            }
+
+            if (isAscii)
+            {
+                char16 *ret = RecyclerNewArrayLeaf(scriptContext->GetRecycler(), char16, UInt32Math::Add(pThisLength, 1));
+                const char16 diffBetweenCases = 32;
+                for (charcount_t i = 0; i < pThisLength; i++)
+                {
+                    char16 cur = pThisSz[i];
+                    if (toUpper)
```

Comment:
Could you lift the check for `toUpper`  outside of the for loop so that you have *less* branching inside the loop?
The less branches, the more the CPU's branch predictor can help you out -- making this loop even _faster_.

Reply:
I would expect that the compiler is clever enough to not emit a run-time check for a template parameter.

---
In reply to: [182906767](https://github.com/Microsoft/ChakraCore/pull/5021#discussion_r182906767) [](ancestors = 182906767)

Reply:
`toUpper` is a template parameter, so this conditional should be compiled away in both `toUpper` and `toLower` cases

Reply:
Ah good point. The small snippet of code that GitHub shows in the diff omitted the template code, which is further up above.